### PR TITLE
Adding an initializer to initialize a model to pretrained weights

### PR DESCRIPTION
**Pull request by danieldeutsch**
This is an initializer that I've been using that initializes some weights to a pretrained model. It addresses #1607. It will be a breaking change because of the additional argument to the initializer's `__call__` method.

 If you all are interested in using this solution, I'm happy to add docs, tests, etc.

### CODE REVIEW

**Comment by matt-gardner** [comment]
This general approach seems reasonable to me, though there are a few details I'd quibble with (I can give those later).  @joelgrus, @DeNeutoy, anyone else: any thoughts?

**Comment by matt-gardner** [comment]
Well, one major problem: if you're initializing many pieces of your model using this, and you're using `from_params`, you're going to have to have many copies of this initializer hanging around, with how we currently have things set up.

**Comment by danieldeutsch** [comment]
Ok, then there might be some details that I don't understand yet. I don't follow the `from_params` part of your comment.

As far as I can tell, you would only need 1 copy of this initializer per InitializerApplicator and per weights file. As it's written now, you can initialize multiple parameters of the model with just 1 instance.

**Comment by DeNeutoy** [comment]
I also don't see the problem - I think this is a great change, super clean approach and achieves something that is otherwise pretty difficult 👍

**Comment by joelgrus** [comment]
possibly what mattg is worried about is e.g. if you need to load several layers from a large pretrained model -- the naive approach would load that model several times, but most likely if you write a nasty enough regex you can only load it once? 😀

**Comment by matt-gardner** [comment]
Yeah, the problem is that each entry for the `InitializerApplication` instantiates its own `Initializer`, so this would get instantiated multiple times, unless you have a really nasty regex.  But, looks like there's general agreement that this is a good idea, so @danieldeutsch, go ahead and add tests and docs and such.

**Comment by joelgrus** [comment]
I will say that I'm not a fan of the name `name_overrides`, it took me a long time to figure out what that was doing -- basically it's a mapping from `name_in_current_model` to `weight_in_pretrained_model`? In that case I'd prefer a name like `name_weight_mapping` or something else that is more descriptive about what it's for

**Comment by matt-gardner** [comment]
Yeah, I have a few other nits too, but I'll worry about those later.

**Comment by danieldeutsch** [comment]
Yes @joelgrus, that's what the `name_overrides` is doing. I chose that name because I wanted it to be clear that it's optional and only necessary when you need to change the name of the parameters. My particular use case was that I trained a model with MLE and then fine tuned with a minimum risk loss function, so none of the parameters' names changed. `name_weight_mapping` doesn't sound optional to me, but I will think about it more.

I'll go ahead and clean it up, then update the PR when it's done

**Comment by matt-gardner** [comment]
Looks like most of the test failures are where we've hacked a constant initializer in test code.  Either we can move all of those to using the actual `constant` initializer (probably cleaner), or we can add the extra parameter to the lambda functions.

There's also a failure because of a hack to get the initializers to work in the config explorer.  You can probably just add another special case for your new initializer [here](https://github.com/allenai/allennlp/blob/e3e8e1cb99d9fc6047070c55dfaf7a2c09316612/allennlp/common/configuration.py#L41).

File: allennlp/nn/initializers.py [hunk 1]
```diff
@@ -198,6 +200,83 @@ def from_params(cls, params: Params):  # type: ignore
...
+                                     "Found %s and %s" % (parameter_name,
+                                                          tensor.data.size(),
+                                                          source_weights.size()))
+
+        # Copy the parameters from the source to the destination
+        tensor.data[:] = source_weights[:]
+
+    @classmethod
+    def from_params(cls, params: Params) -> 'PretrainedModelInitializer':  # type: ignore
+        # The ``from_params`` method has to be defined, otherwise the
```
[the remark this answers is not in the document]
Comment (danieldeutsch) [reply allennlp/nn/initializers.py]: I can remove both.

If I did not define my own `from_params`, `Initializer`'s is called, it pops the type "pretrained" and calls `PretrainedModelInitializer`'s method, which does not exist, so it actually calls itself. Then, when `cls` is `PretrainedModelInitializer`, before it pops the type from the params, it first calls `cls.list_available()`. Within this method, it checks to ensure that the default implementation is registered. Since the default implementation for `PretrainedModelInitializer` does not exist, it uses `Initializer`, which is "normal". But because this is not registered for `PretrainedModelInitializer`, it raises a ConfigurationError [here](https://github.com/allenai/allennlp/blob/193bb04ddae7d89ee63005b898a01641447d6074/allennlp/common/registrable.py#L71).

File: allennlp/tests/nn/pretrained_model_initializer_test.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        self.linear_1 = torch.nn.Linear(5, 10)
+        self.linear_3 = torch.nn.Linear(10, 5)
+
+    def forward(self, inputs):  # pylint: disable=arguments-differ
+        pass
+
+
+class TestPretrainedModelInitializer(AllenNlpTestCase):
+    def setUp(self):
+        super(TestPretrainedModelInitializer, self).setUp()
```
[the remark this answers is not in the document]
Comment (danieldeutsch) [reply allennlp/tests/nn/pretrained_model_initializer_test.py]: I copied this from `initializers_test.py`, but I will remove it. Should I remove the ones from the `__init__`` methods too?

File: allennlp/nn/initializers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/nn/initializers.py]: Ah, that's a really interesting side effect of how we do `FromParams` that I hadn't considered - if we ever have a `Registrable` class that implements its own `from_params`, _all_ subclasses must also implement it.  That's unfortunate...

File: allennlp/tests/nn/pretrained_model_initializer_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/tests/nn/pretrained_model_initializer_test.py]: Yes please.  Those arguments are almost never needed.  I used to add them (and some others do too), thinking they were necessary, but they aren't.

File: allennlp/nn/initializers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelgrus) [reply allennlp/nn/initializers.py]: I wouldn't call that a "side effect" per se -- the automatic `FromParams` is defined in the deepest base class, so once it's overridden somewhere in a class hierarchy, it's overridden the rest of the way up.

But presumably we could add logic to `Initializer.from_params` that checks if `cls == Initializer` and if not it calls `FromParams.from_params(cls, params)`, I think that would work? Maybe it's too hacky?

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/nn/initializers.py]: I think that's a good solution for future cases where we have something like this.  For `Initializer`, it's not actually necessary, because we can just remove the method entirely.

**Comment by danieldeutsch** [comment]
I addressed your comments, and I'm fixing the pylint issue now.

I was sort of blindly fixing the unit tests because whenever I ran pytest locally, I kept getting an error that there were too many files open. Have you ever seen this?

**Comment by matt-gardner** [comment]
@danieldeutsch, hmm, yeah, I'll take a look at this.  The rest of my week is pretty crazy, though, so it'll probably be on Monday or Tuesday.

**Comment by matt-gardner** [comment]
@joelgrus, can you give me a quick sanity check that the (two-line) change I made to `FromParams.from_params` is reasonable?  You can see it easiest by clicking on the "handle plain str" commit above.  I think everything else here has been reviewed.
